ext/semian/sysv_semaphores.c

```diff
@@ -80,19 +81,27 @@ set_semaphore_permissions(int sem_id, long permissions)
 }
 
 int
-perform_semop(int sem_id, short index, short op, short flags, struct timespec *ts)
+perform_semop(int sem_id, short index, short op, short flags, struct timespec *ts, int *wait_time_ms)
 {
   struct sembuf buf = { 0 };
 
   buf.sem_num = index;
   buf.sem_op  = op;
   buf.sem_flg = flags;
 
+  struct timespec begin, end;
+  if (wait_time_ms) {
+    clock_gettime(CLOCK_MONOTONIC, &begin);
+  }
   if (ts) {
     return semtimedop(sem_id, &buf, 1, ts);
   } else {
     return semop(sem_id, &buf, 1);
   }
+  if (wait_time_ms) {
+    clock_gettime(CLOCK_MONOTONIC, &end);
+    *wait_time_ms = begin.tv_sec * 1e3 + begin.tv_nsec / 1e6 - end.tv_sec * 1e3 - end.tv_nsec / 1e6;
```

Comment:
This line isn't very descriptive, I'd recommend refactoring to a static function to better describe the nature of this (relatively complicated) timing operation.

Alternatively, if you're concerned about speed, an inline could make sense - but i'd been avoiding them where possible.